src/org/mozilla/javascript/optimizer/BodyCodegen.java

```diff
@@ -3328,112 +3276,145 @@ private void visitIncDec(Node node)
+                    cfw.addALoad(variableObjectLocal);
+                    cfw.addPush(incrDecrMask);
+                    addScriptRuntimeInvoke(
+                            "refIncrDecr",
+                            "(Lorg/mozilla/javascript/Ref;"
+                                    + "Lorg/mozilla/javascript/Context;"
+                                    + "Lorg/mozilla/javascript/Scriptable;"
+                                    + "I)Ljava/lang/Object;");
+                    break;
                 }
-                break;
-            }
-            case Token.GET_REF: {
-                Node refChild = child.getFirstChild();
-                generateExpression(refChild, node);
-                cfw.addALoad(contextLocal);
-                cfw.addALoad(variableObjectLocal);
-                cfw.addPush(incrDecrMask);
-                addScriptRuntimeInvoke(
-                    "refIncrDecr",
-                    "(Lorg/mozilla/javascript/Ref;"
-                        +"Lorg/mozilla/javascript/Context;"
-                        +"Lorg/mozilla/javascript/Scriptable;"
-                        +"I)Ljava/lang/Object;");
-                break;
-            }
             default:
                 Codegen.badTree();
         }
     }
 
-    private static boolean isArithmeticNode(Node node)
-    {
+    private static boolean isArithmeticNode(Node node) {
         int type = node.getType();
         return (type == Token.SUB)
-            || (type == Token.MOD)
-            || (type == Token.DIV)
-            || (type == Token.MUL);
+                || (type == Token.MOD)
+                || (type == Token.DIV)
+                || (type == Token.MUL);
     }
 
-    private void visitArithmetic(Node node, int opCode, Node child,
-        Node parent)
-    {
+    private void visitArithmetic(Node node, int type, Node child, Node parent) {
         int childNumberFlag = node.getIntProp(Node.ISNUMBER_PROP, -1);
         if (childNumberFlag != -1) {
             generateExpression(child, node);
             generateExpression(child.getNext(), node);
-            cfw.add(opCode);
-        }
-        else {
-            boolean childOfArithmetic = isArithmeticNode(parent);
+
+            switch (type) {
+                case Token.SUB:
+                    cfw.add(ByteCode.DSUB);
+                    break;
+                case Token.MUL:
+                    cfw.add(ByteCode.DMUL);
+                    break;
+                case Token.DIV:
+                    cfw.add(ByteCode.DDIV);
+                    break;
+                case Token.MOD:
+                    cfw.add(ByteCode.DREM);
+                    break;
+                default:
+                    throw Kit.codeBug(Token.typeToName(type));
+            }
+        } else {
             generateExpression(child, node);
-            if (!isArithmeticNode(child))
-                addObjectToDouble();
+            if (!isArithmeticNode(child)) addObjectToNumeric();
```

Comment:
On optimization with `isArithmeticNode`:

1. This PR
2. Make the condition of `isArithmeticNode` the same as this
    https://github.com/mozilla/rhino/blob/0c05d30e134aed65b9c3536d8ac3764590840a35/src/org/mozilla/javascript/optimizer/Block.java#L482-L498
3. Delete `isArithmeticNode`

When I compared these results, `./gradlew testBenchmark` results were almost the same.
This is just a guess, but this benchmark was run with `OptimizationLevel` of 9, so it seems to be optimized by `childNumberFlag`. We may be able to find a way to speed up the process by setting the `OptimizationLevel` to 0 and verifying it.